### PR TITLE
Support for PyTorch based models with TorchModel

### LINKED ISSUE #2059: Design of PyTorch models
Now that we're considering adding models based on PyTorch, it seems like a good time to consider the design for them.  We'll want to minimize code duplication, and also ensure the API is as consistent as possible (both between PyTorch based models, and between them and other DeepChem models).  At the same time, we should maintain lots of flexibility and make it easy to use existing PyTorch models from outside.

To get the discussion going, let me offer a design proposal.  This is intended just as a starting point for discussion, so please feel free to suggest changes or alternate designs.

We will create a TorchModel class whose API and feature set are as close as possible to KerasModel.  In place of a `tf.keras.Model`, the constructor will take a `torch.nn.Module`.  Aside from that one difference, the API will be nearly identical.  It will use the same Optimizer, Loss, and Callback objects.  I'm not sure whether all features of KerasModel can be supported, but as far as possible we'll try to do it.

Would that work?  Are there any serious problems we would run into with that approach?

### PR DESCRIPTION
**Pull request by peastman**
I still have some features left to implement and some cleanup to do, but this is largely functional.  The API is nearly identical to KerasModel.  For a lot of the test cases, converting them involved nothing more than replacing

```python
  keras_model = tf.keras.Sequential([
      tf.keras.layers.Dense(10, activation='relu'),
      tf.keras.layers.Dense(1, activation='sigmoid')
  ])
```

with

```python
  pytorch_model = torch.nn.Sequential(
      torch.nn.Linear(2, 10), torch.nn.ReLU(), torch.nn.Linear(10, 1),
      torch.nn.Sigmoid())
```

How do we want to handle installing PyTorch?  One option is to add it to requirements.yml so it will automatically be installed whenever someone installs DeepChem.  That would probably be the simplest approach, but it would add another very large download for every user.  Alternatively we could leave it as optional and install it separately on Travis.

Fixes #2059.

### CODE REVIEW

File: deepchem/models/tests/test_losses.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  has_tensorflow = False
+
+try:
+  import torch
+  has_pytorch = True
+except:
+  has_pytorch = False
+
+
+class TestLosses(unittest.TestCase):
```
Comment (rbharath) [inline deepchem/models/tests/test_losses.py:18]: It might be good to convert these tests to the new pytest style (functions without the wrapper `unittest.TestCase` class

File: deepchem/models/tests/test_losses.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  """Test loss functions."""
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
```
Comment (rbharath) [inline deepchem/models/tests/test_losses.py:21]: I think `pytest.skip` provides equivalent functionality

File: deepchem/models/tests/test_optimizers.py [hunk 3]
```diff
@@ -1,57 +1,137 @@
...
 
-class TestLayers(test_util.TensorFlowTestCase):
+try:
+  import torch
+  has_pytorch = True
+except:
+  has_pytorch = False
+
+
+class TestOptimizers(unittest.TestCase):
```
Comment (rbharath) [inline deepchem/models/tests/test_optimizers.py:17]: Same suggestion here

File: deepchem/utils/typing.py [hunk 4]
```diff
@@ -7,8 +7,8 @@
 # An activation function for a Keras layer: either a TensorFlow function or the name of a standard activation
 KerasActivationFn = Union[Callable, str]
```
Comment (rbharath) [inline deepchem/utils/typing.py:8]: Should we rename the other parameters to remove `Keras` (eg. `KerasActivationFn` -> `ActivationFn`?

File: deepchem/models/torch_model.py [hunk 5]
```diff
@@ -0,0 +1,1167 @@
...
+from deepchem.models.models import Model
+from deepchem.models.optimizers import Adam, Optimizer, LearningRateSchedule
+from deepchem.trans import Transformer, undo_transforms
+from deepchem.utils.evaluate import GeneratorEvaluator
+
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple, Union
+from deepchem.utils.typing import LossFn, OneOrMany
+
+try:
+  import wandb
```
Comment (rbharath) [inline deepchem/models/torch_model.py:26]: Nice we have wandb support added out of box :)

File: deepchem/models/torch_model.py [hunk 6]
```diff
@@ -0,0 +1,1167 @@
...
+    _has_wandb = True
+except (ImportError, AttributeError):
+  _has_wandb = False
+
+
+def is_wandb_available():
+  return _has_wandb
+
+
+class TorchModel(Model):
```
Comment (rbharath) [inline deepchem/models/torch_model.py:43]: This might be out of the scope of this PR, but it might be really useful to add a page to `deepchem/docs` that explains how to develop custom models in DeepChem with Keras and PyTorch examples. Now that our infrastructure is maturing, we're likely to see more folks who use DeepChem for custom applications so a guide might be really useful

File: deepchem/models/torch_model.py [hunk 7]
```diff
@@ -0,0 +1,1167 @@
...
+    must have the same shape as the corresponding prediction
+    output, and each element is an estimate of the variance in
+    the corresponding prediction.  Also be aware that if a model
+    supports uncertainty, it MUST use dropout on every layer,
+    and dropout most be enabled during uncertainty prediction.
+    Otherwise, the uncertainties it computes will be inaccurate.
+
+  - other: Arbitrary output_types can be used to extract outputs
+    produced by the model, but will have no additional
+    processing performed.
```
Comment (rbharath) [inline deepchem/models/torch_model.py:102]: It might be useful to add an examples section to the docstring demonstrating how to use TorchModel on a simple toy model

File: deepchem/models/tests/test_losses.py (hunk 2 above)
Comment (peastman) [reply deepchem/models/tests/test_losses.py:21]: I prefer to use the `unittest` versions because it's built in and supported everywhere.  When we switched to pytest, everything specific to nosetest had to be converted, but everything using unittest kept working.  Who knows if we'll switch again in the future?

File: deepchem/utils/typing.py (hunk 4 above)
Comment (peastman) [reply deepchem/utils/typing.py:8]: This one is specific to Keras.  It refers to how layers let you write either `activation=tf.nn.relu` or `activation='relu'`.

File: deepchem/models/torch_model.py (hunk 6 above)
Comment (peastman) [reply deepchem/models/torch_model.py:43]: Yes, definitely!  (But that's a separate PR.)

**Comment by peastman** [comment]
> My two cents is that we should leave PyTorch as an optional dependency.

Ok, I'll add it to travis then.

**Comment by peastman** [comment]
It appears that Travis is already installing PyTorch, though I don't see anywhere that we tell it to.  Is it a dependency of something else we install?

**Comment by nd-02110114** [comment]
Now, Travis install PyTorch through `scripts/install_deepchem_conda.sh`.
https://github.com/deepchem/deepchem/blob/8217de5ab76a7cd62d0524471b8b3569601f109a/scripts/install_deepchem_conda.sh#L56

**Comment by peastman** [comment]
This is ready for a full review.  All features are now working.  The only feature of KerasModel that isn't supported in TorchModel is specifying a custom set of outputs for `predict()`.

File: deepchem/models/torch_model.py (hunk 7 above)
Comment (peastman) [reply deepchem/models/torch_model.py:110]: Added!  (Also for KerasModel.)

File: deepchem/models/torch_model.py [hunk 8]
```diff
@@ -0,0 +1,1124 @@
...
+                     batch: Tuple[Any, Any, Any]) -> Tuple[List, List, List]:
+    inputs, labels, weights = batch
+    inputs = [
+        x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs
+    ]
+    if labels is not None:
+      labels = [
+          x.astype(np.float32) if x.dtype == np.float64 else x for x in labels
+      ]
+      labels = [torch.as_tensor(x) for x in labels]
```
Comment (nd-02110114) [inline deepchem/models/torch_model.py:837]: I think we should set a value to the device argument.


```python
def get_device():
    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
    return device

device = get_device()
torch.as_tensor(a, device=device)
```

File: deepchem/models/torch_model.py [hunk 9]
```diff
@@ -0,0 +1,1124 @@
...
+      inputs, labels, weights = batch
+      inputs, _, _ = self._prepare_batch((inputs, None, None))
+
+      # Invoke the model.
+      if len(inputs) == 1:
+        inputs = inputs[0]
+      output_values = self.model(inputs)
+      if isinstance(output_values, torch.Tensor):
+        output_values = [output_values]
+      output_values = [t.detach().numpy() for t in output_values]
```
Comment (nd-02110114) [inline deepchem/models/torch_model.py:525]: I think this code will not work with GPU.

If we use GPU, we call `cpu()` before calling `numpy()`.

```python
def tensor2numpy(value):
    if torch.cuda.is_available():
        return value.detach().cpu().numpy()
    else:
        return value.detach().numpy()

output_values = [tensor2numpy(t) for t in output_values]

```

File: deepchem/models/torch_model.py [hunk 10]
```diff
@@ -0,0 +1,1124 @@
...
+    Parameters
+    ----------
+    source_model: dc.models.TorchModel
+        Source model to create value map from
+    """
+    value_map: Dict[Any, Any] = {}
+    source_vars = list(source_model.model.parameters())
+
+    for source_var in source_vars:
+      value_map[source_var] = source_var.detach().numpy()
```
Comment (nd-02110114) [inline deepchem/models/torch_model.py:1027]: Same here

File: deepchem/models/losses.py [hunk 11]
```diff
@@ -25,24 +22,38 @@ def __call__(self, output, labels):
     """
     raise NotImplementedError("Subclasses must implement this")
 
+  def _create_pytorch_loss(self):
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/losses.py:25]: Losses work differently in TensorFlow and PyTorch.  For TensorFlow, `_compute_tf_loss()` directly computes the loss function, so it takes outputs and labels.  For PyTorch, it creates and returns an object that computes the loss function.  `_create_pytorch_loss()` doesn't take any arguments because it doesn't need them.  The object it returns is callable and takes outputs and labels as arguments.

File: deepchem/models/losses.py [hunk 12]
```diff
@@ -91,11 +130,21 @@ class SigmoidCrossEntropy(Loss):
...
-  def __call__(self, output, labels):
+  def _compute_tf_loss(self, output, labels):
+    import tensorflow as tf
     output, labels = _make_shapes_consistent(output, labels)
     output, labels = _ensure_float(output, labels)
     return tf.nn.sigmoid_cross_entropy_with_logits(labels, output)
 
+  def _create_pytorch_loss(self):
+    import torch
+    bce = torch.nn.BCELoss(reduction='none')
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/losses.py:141]: Good suggestion, thanks.

File: deepchem/models/losses.py [hunk 13]
```diff
@@ -106,11 +155,21 @@ class SoftmaxCrossEntropy(Loss):
...
 
-  def __call__(self, output, labels):
+  def _compute_tf_loss(self, output, labels):
+    import tensorflow as tf
     output, labels = _make_shapes_consistent(output, labels)
     output, labels = _ensure_float(output, labels)
     return tf.nn.softmax_cross_entropy_with_logits(labels, output)
 
+  def _create_pytorch_loss(self):
+    import torch
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/losses.py:165]: `CrossEntropyLoss` requires the labels to be a class index, but we get them as one-hot encoded instead.

File: deepchem/models/losses.py [hunk 14]
```diff
@@ -77,11 +107,20 @@ class CategoricalCrossEntropy(Loss):
...
 
-  def __call__(self, output, labels):
+  def _compute_tf_loss(self, output, labels):
+    import tensorflow as tf
     output, labels = _make_shapes_consistent(output, labels)
     output, labels = _ensure_float(output, labels)
     return tf.keras.losses.categorical_crossentropy(labels, output)
 
+  def _create_pytorch_loss(self):
+    import torch
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/losses.py:117]: Same here.

File: deepchem/models/torch_model.py (hunk 8 above)
Comment (peastman) [reply deepchem/models/torch_model.py:837]: Good suggestion.  I haven't tested any of this on a GPU yet.  That's next up on my list.

File: deepchem/models/tests/test_losses.py (hunk 2 above)
Comment (rbharath) [reply deepchem/models/tests/test_losses.py:21]: Good point!

File: deepchem/models/losses.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/losses.py:165]: I changed it to use `torch.nn.LogSoftmax`.

File: deepchem/models/tests/test_losses.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+    result = loss._create_pytorch_loss()(outputs, labels).numpy()
+    expected = [np.mean([0.9, 1.8]), np.mean([1.4, 0.4])]
+    assert np.allclose(expected, result)
+
+  @unittest.skipIf(not has_tensorflow, 'TensorFlow is not installed')
+  def test_binary_cross_entropy_tf(self):
+    """Test BinaryCrossEntropy."""
+    loss = losses.BinaryCrossEntropy()
+    outputs = tf.constant([[0.1, 0.8], [0.4, 0.6]])
+    labels = torch.tensor([[0.0, 1.0], [1.0, 0.0]])
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/tests/test_losses.py:86]: Why?  It works fine.

```
>>> torch.tensor([[0.0, 1.0], [1.0, 0.0]])
tensor([[0., 1.],
        [1., 0.]])
```

File: deepchem/models/optimizers.py [hunk 16]
```diff
@@ -23,14 +35,28 @@ def _create_optimizer(self, global_step):
...
+
 
 class LearningRateSchedule(object):
   """A schedule for changing the learning rate over the course of optimization.
 
   This is an abstract class.  Subclasses represent specific schedules.
   """
 
-  def _create_tensor(self, global_step):
+  def _create_tf_tensor(self, global_step):
```
[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/optimizers.py:59]: The names refer to differences in how learning rates work in the two packages.  The TensorFlow version returns a tensor that holds the current learning rate (and gets updated automatically).  The PyTorch version returns a learning rate scheduler that you invoke to update the optimizer's learning rate.

File: deepchem/models/tests/test_losses.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nd-02110114) [reply deepchem/models/tests/test_losses.py:86]: In general, do we calculate loss values between torch.Tensor and tf.Tensor? I think this answer is No. If we use tensorflow, labels and outputs are also defined by tf.Tensor. So, I think it is better to define labels using tf.Tensor compared with torch.Tensor. 

And, the test name is `test_binary_cross_entropy_tf`.
I seem that `XXX_tf` tests expect only tensorflow dependency. (`unittest.skipIf` also implify this) So, I think it is not better to depend on PyTorch.  (If you think you really want to use PyTorch in order to define labels using torch.Tensor, I think you should also add `@unittest.skipIf(not has_pytorch, 'PyTorch is not installed')`)

[the remark this answers is not in the document]
Comment (peastman) [reply deepchem/models/tests/test_losses.py:86]: Oh, I see!  I though you were saying it should be `torch.constant()` instead of `torch.tensor()`.  Yes, this was an accident.  TensorFlow tests should use TensorFlow classes and PyTorch tests should use PyTorch classes.

**Comment by nd-02110114** [comment]
LGTM

**Comment by rbharath** [comment]
Going to go ahead and merge this in!
